### PR TITLE
Real file publishing

### PR DESCRIPTION
**Pull request by philipandag**
Publish file now makes the node only provide the file, it does not publish it in the network. This needs to change.
I will:
- change the name of current PublishFile implementation to ProvideFile
- Add actual implementation of file publishing

### CODE REVIEW

**Comment by philipandag** [comment]
I have cleaned up the humongous matches enforced by the libp2p events by moving the arm handlers into functions. It don't think it is much better, but a bit at least and I didn't have a better idea at that moment. At least the functions can now be documented. Another option would be to move them to modules but I am not sure if that would help much.

The option to publish the file to DHT was removed. The `PutRecord` kademlia message is now used only as a request to start providing the file using the `request_response` behavior. Implementing a Vault into this should not be that hard. The places where using the Vault makes sense are marked with comments.

I highly advocate to merge this PR because it blocks pretty much any further development.

**Comment by philipandag** [comment]
I have added some documentation comments that might hopefully help with wrapping our heads around this code, which is not that complex really. It's just the events framework that libp2p proposes that makes it hard to track the flow of the execution.
Maybe generating docs could help?
`cargo doc --no-deps --document-private-items --bins --open`

File: liberum_core/src/swarm_runner/behaviour/kademlia.rs [hunk 1]
```diff
@@ -1,87 +1,353 @@
...
+                node = self.node.name,
+                err = format!("{qid:?}"),
+                "Failed to start providing file"
+            );
+            return;
+        }
+        let qid = qid.unwrap();
+
+        // We can't await for a response here because it would block the event loop
+        let (response_sender, _) = oneshot::channel();
```
Comment (originalmk) [inline liberum_core/src/swarm_runner/behaviour/kademlia.rs:272]: We are dropping receiver here, don't we want to get an answer?

File: liberum_core/src/swarm_runner/behaviour/kademlia.rs [hunk 2]
```diff
@@ -1,87 +1,353 @@
...
+
+    fn handle_inbound_request_add_provider(&mut self, record: Option<ProviderRecord>) {
+        match record {
+            Some(record) => {
+                self.swarm
+                    .behaviour_mut()
+                    .kademlia
+                    .store_mut()
+                    .add_provider(record.clone())
+                    .ok(); // TODO What if the providers amount is exceeded? How to ensure only the closest one are kept?
```
Comment (originalmk) [inline liberum_core/src/swarm_runner/behaviour/kademlia.rs:286]: I believe that GC was a solution to this problem, so cleaning up all the provider records every 24 hours and requiring providers to resend info about provided files across the network.

File: liberum_core/src/swarm_runner/behaviour/kademlia.rs [hunk 3]
```diff
@@ -1,87 +1,353 @@
...
+            }
+        }
+    }
+
+    fn handle_inbound_request_get_provider(
+        &mut self,
+        _num_closer_peers: usize,
+        _num_provider_peers: usize,
+    ) {
+        debug!(node = self.node.name, "Kad Received GetProvider")
```
Comment (originalmk) [inline liberum_core/src/swarm_runner/behaviour/kademlia.rs:306]: Don't we want to handle that request?

File: liberum_core/src/swarm_runner/messages.rs [hunk 4]
```diff
@@ -92,22 +118,22 @@ impl SwarmContext {
...
                     );
                     return Ok(false);
                 }
-                // Add the file to the published list
-                self.behaviour.published.insert(
+                // Add the file to the providing list
+                self.behaviour.providing.insert(
                     id.clone(),
                     file_share::SharedResource::File { path: path.clone() },
                 );
```
Comment (originalmk) [inline liberum_core/src/swarm_runner/messages.rs:139]: It is not part of this PR, but why do we want to send GetProviders while handling ProvideFile message?

File: liberum_core/src/swarm_runner/messages.rs [hunk 5]
```diff
@@ -134,18 +160,65 @@ impl SwarmContext {
...
+                //self.swarm.behaviour_mut().kademlia.start_providing(record.key.clone())?;
+                //self.put_record_into_vault(record.clone());
+
                 let qid = self
                     .swarm
                     .behaviour_mut()
-                    .file_share
-                    .send_request(&peer, file_share::FileRequest { id: id.to_vec() });
+                    .kademlia
+                    .put_record(record, kad::Quorum::One)?; // at least one other node MUST store to consider it successful
```
Comment (originalmk) [inline liberum_core/src/swarm_runner/messages.rs:218]: One may be too low I think...

**Review by originalmk — CHANGES_REQUESTED** [review]

**Comment by originalmk** [comment]
I haven't reviewed the tests, but I believe they are working right.

File: liberum_cli/src/main.rs [hunk 6]
```diff
@@ -402,7 +412,25 @@ async fn handle_get_providers(
...
 
-    handle_response(&mut res).await
+    let response = res
+        .recv()
+        .await
+        .ok_or(anyhow!("Daemon returned no response"))?;
+    match response {
+        Ok(DaemonResponse::Providers { ids }) => {
+            for provider in ids {
+                println!("{provider}");
```
[the remark this answers is not in the document]
Comment (philipandag) [reply liberum_cli/src/main.rs:422]: Yes, the CLI should give the list of providers when you ask it for providers

File: liberum_cli/src/main.rs [hunk 7]
```diff
@@ -445,6 +473,37 @@ async fn handle_dial(cmd: Dial, req: RequestSender, mut res: ReseponseReceiver)
...
+    mut res: ReseponseReceiver,
+) -> Result<()> {
+    req.send(DaemonRequest::PublishFile {
+        node_name: cmd.node_name,
+        path: cmd.path,
+    })
+    .await
+    .inspect_err(|e| error!(err = e.to_string(), "Failed to send message"))?;
+
+    let resp = res
```
[the remark this answers is not in the document]
Comment (philipandag) [reply liberum_cli/src/main.rs:488]: It prints a meaningful text instead of a debug print of the message.

File: liberum_core/src/connection/mod.rs [hunk 8]
```diff
@@ -52,7 +53,7 @@ pub async fn listen(listener: UnixListener) -> Result<()> {
     let app_context = AppContext::new(kameo::spawn(NodeStore::with_default_nodes_dir().await?));
     loop {
         let (daemon_socket, _) = listener.accept().await?;
-        info!(conn_id = id, "Handling a new connection");
+        //info!(conn_id = id, "Handling a new connection");
```
[the remark this answers is not in the document]
Comment (philipandag) [reply liberum_core/src/connection/mod.rs:56]: I wasn't sure if it should be removed. It is printed every time a client connects to the core, which happens a lot when using the CLI and isn't very informational once we know that this basic operation works

File: liberum_core/src/node/mod.rs [hunk 9]
```diff
@@ -203,6 +224,52 @@ impl Node {
         }
         Err(anyhow!("Swarm sender is None"))
     }
+
+    #[message]
+    pub async fn publish_file(&mut self, path: PathBuf) -> Result<String> {
+        if let Some(sender) = &mut self.swarm_sender {
+            let id = liberum_core::get_file_id(&path).await.map_err(|e| {
```
[the remark this answers is not in the document]
Comment (philipandag) [reply liberum_core/src/node/mod.rs]: https://github.com/Liberum-Neto/liberum-neto/pull/43/commits/fb6d8e674ce47ae8bc9eb33311806e566c3ae9e6

File: liberum_core/src/node/mod.rs [hunk 10]
```diff
@@ -203,6 +224,52 @@ impl Node {
...
+                    "Failed to hash file"
+                );
+                e
+            })?;
+            let (send, recv) = oneshot::channel();
+
+            // The file has to be read to the memory to be published. There is no other way without
+            // a new behaviour kademlia could talk to, which would provide streams of data.
+            // (Maybe could be implemented on the existing request_response if it would be generalised more?)
+            let data = tokio::fs::read(&path).await.map_err(|e| {
```
[the remark this answers is not in the document]
Comment (philipandag) [reply liberum_core/src/node/mod.rs]: https://github.com/Liberum-Neto/liberum-neto/pull/43/commits/fb6d8e674ce47ae8bc9eb33311806e566c3ae9e6

File: liberum_core/src/swarm_runner/behaviour/file_share.rs [hunk 11]
```diff
@@ -29,51 +38,76 @@ impl SwarmContext {
...
+/// Methods on SwarmContext for handling file sharing
+impl SwarmContext {
+    /// Handle a file share request depending on the type of the data which ID is requested
+    async fn handle_file_share_request(
+        &mut self,
+        _request_id: InboundRequestId,
+        request: FileRequest,
+        response_channel: ResponseChannel<FileResponse>,
+    ) {
+        debug!("Request_response request!");
```
[the remark this answers is not in the document]
Comment (philipandag) [reply liberum_core/src/swarm_runner/behaviour/file_share.rs]: https://github.com/Liberum-Neto/liberum-neto/pull/43/commits/cda15daa7ec768396b4de9347c3fb8a8b8e0ff14

File: liberum_core/src/swarm_runner/behaviour/file_share.rs [hunk 12]
```diff
@@ -29,51 +38,76 @@ impl SwarmContext {
...
+        // Get the file from the providing hashmap
+        let id = kad::RecordKey::from(request.id.clone());
+        let file = self.behaviour.providing.get(&id);
+        // Send the file back to the peer if found
+        if let Some(file) = file {
```
[the remark this answers is not in the document]
Comment (philipandag) [reply liberum_core/src/swarm_runner/behaviour/file_share.rs:77]: It should be fixed. Probably on both of the sides. The requested node should return an Err or None and the requesting one should implement some timeout.

[the remark this answers is not in the document]
Comment (philipandag) [reply liberum_core/src/swarm_runner/behaviour/file_share.rs:80]: https://github.com/Liberum-Neto/liberum-neto/pull/43/commits/bcb9fa1a43531fbf405a8b0e1db198c06918a926

File: liberum_core/src/connection/mod.rs [hunk 13]
```diff
@@ -82,7 +83,7 @@ async fn handle_connection(
                 };
             },
             else => {
-                debug!(conn_id=id, "Connection closed");
+                //debug!(conn_id=id, "Connection closed");
```
[the remark this answers is not in the document]
Comment (philipandag) [reply liberum_core/src/connection/mod.rs:86]: same as above, I don't think it's that useful to log every time a client closes its connection. Should it be removed?

File: liberum_core/src/swarm_runner/behaviour/file_share.rs [hunk 14]
```diff
@@ -29,51 +38,76 @@ impl SwarmContext {
...
+        }
+    }
+
+    /// Handle a file share response by sending the data to the pending download
+    fn handle_file_share_response(
+        &mut self,
+        request_id: OutboundRequestId,
+        response: FileResponse,
+    ) {
+        debug!("Request_response response!");
```
[the remark this answers is not in the document]
Comment (philipandag) [reply liberum_core/src/swarm_runner/behaviour/file_share.rs]: https://github.com/Liberum-Neto/liberum-neto/commit/cda15daa7ec768396b4de9347c3fb8a8b8e0ff14

File: liberum_core/src/swarm_runner/behaviour/kademlia.rs [hunk 15]
```diff
@@ -1,87 +1,353 @@
...
+    fn handle_outbound_query_progressed_start_providing(
+        &mut self,
+        id: QueryId,
+        result: Result<AddProviderOk, AddProviderError>,
+        _stats: QueryStats,
+        _step: ProgressStep,
+    ) {
+        // result is matched two times to trace the result no matter
+        // if the response is sent via the oneshot channel or not
+        if result.is_ok() {
```
[the remark this answers is not in the document]
Comment (philipandag) [reply liberum_core/src/swarm_runner/behaviour/kademlia.rs:108]: It would require cloning the result as if let will take it's posession

File: liberum_core/src/swarm_runner/behaviour/kademlia.rs [hunk 16]
```diff
@@ -1,87 +1,353 @@
...
+
+        // Respond to the caller of StartProviding
+        let sender = self.behaviour.pending_start_providing.remove(&id);
+        if let Some(sender) = sender {
+            match result {
+                Ok(_) => {
+                    let _ = sender.send(Ok(()));
+                }
+                Err(e) => {
+                    let _ = sender.send(Err(anyhow!(e)));
```
[the remark this answers is not in the document]
Comment (philipandag) [reply liberum_core/src/swarm_runner/behaviour/kademlia.rs]: https://github.com/Liberum-Neto/liberum-neto/pull/43/commits/49a5748e473e095552a8bb4ee948e6efd9366d0b

File: liberum_core/src/swarm_runner/behaviour/kademlia.rs [hunk 17]
```diff
@@ -1,87 +1,353 @@
...
+
+        let sender = self.behaviour.pending_publish_file.remove(&id);
+
+        if let Some(sender) = sender {
+            match result {
+                Ok(_) => {
+                    let _ = sender.send(Ok(()));
+                }
+                Err(e) => {
+                    let _ = sender.send(Err(anyhow!(e)));
```
[the remark this answers is not in the document]
Comment (philipandag) [reply liberum_core/src/swarm_runner/behaviour/kademlia.rs]: https://github.com/Liberum-Neto/liberum-neto/pull/43/commits/49a5748e473e095552a8bb4ee948e6efd9366d0b

File: liberum_core/src/swarm_runner/behaviour/kademlia.rs [hunk 18]
```diff
@@ -1,87 +1,353 @@
...
+        id: QueryId,
+        _closest_peers: Vec<PeerId>,
+        _stats: QueryStats,
+        _step: ProgressStep,
+    ) {
+        debug!(
+            node = self.node.name,
+            "Get providers didn't find any new records"
+        );
+        if let Some(sender) = self.behaviour.pending_get_providers.remove(&id) {
```
[the remark this answers is not in the document]
Comment (philipandag) [reply liberum_core/src/swarm_runner/behaviour/kademlia.rs:223]: I don't think the query is still in progress once "FinishedWithNoAdditionalRecords" is triggered

File: liberum_core/src/swarm_runner/behaviour/kademlia.rs [hunk 19]
```diff
@@ -1,87 +1,353 @@
...
+            .start_providing(id.clone());
+        if qid.is_err() {
+            debug!(
+                node = self.node.name,
+                err = format!("{qid:?}"),
+                "Failed to start providing file"
+            );
+            return;
+        }
+        let qid = qid.unwrap();
```
[the remark this answers is not in the document]
Comment (philipandag) [reply liberum_core/src/swarm_runner/behaviour/kademlia.rs:264]: It is not possible to return out of function from inside a closure

File: liberum_core/src/swarm_runner/behaviour/kademlia.rs (hunk 1 above)
Comment (philipandag) [reply liberum_core/src/swarm_runner/behaviour/kademlia.rs:267]: We cannot because the answer can only be received from an event handler, which won't be called if the event loop is waiting. A rework of the swarm logic should probably be made to use Arc Mutii and spawn tasks to handle the events instead of blocking the event loop to handle every event.

File: liberum_core/src/swarm_runner/behaviour/kademlia.rs (hunk 2 above)
Comment (philipandag) [reply liberum_core/src/swarm_runner/behaviour/kademlia.rs:281]: There is a hard cap on how many providers the DHT will store (n=K=20 by default). This handler always adds a provider to the DHT when someone informs that they are a provider. I am not sure what happens when the cap is exceeded. Are the oldest ones dropped or what? Probably it's not that crucial...

File: liberum_core/src/swarm_runner/behaviour/kademlia.rs (hunk 3 above)
Comment (philipandag) [reply liberum_core/src/swarm_runner/behaviour/kademlia.rs:301]: There is no need to do anything more here other than logging. This single trace might be droppped too, but I figured it would be helpful for me in the debugging process.

File: liberum_core/src/swarm_runner/messages.rs (hunk 4 above)
Comment (philipandag) [reply liberum_core/src/swarm_runner/messages.rs:139]: I don't know what do you mean. GetProviders is not called. The `providing` hashmap is just a crude implementation of some kind of file store. It has a record <ID, PATH> for every file the Node is providing.

File: liberum_core/src/swarm_runner/messages.rs (hunk 5 above)
Comment (philipandag) [reply liberum_core/src/swarm_runner/messages.rs:218]: That's the absolute minimum of other nodes that should accept the request to provide a file to call it a success. If there are only two nodes in the network a quorum of anything more than one would fail. It's not like only one node will be asked to provide. up to K=20 nodes will be asked to do that. The quorum only decides if the query will be treated as a success or a failure. In a mature network there probably would be no reason to leave it at One though? At this moment it needs to stay low for the tests to work.

File: liberum_core/src/swarm_runner/behaviour/kademlia.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (philipandag) [reply liberum_core/src/swarm_runner/behaviour/kademlia.rs:264]: https://github.com/Liberum-Neto/liberum-neto/pull/43/commits/a484a3d7d72fb02561865bd34086cc63cf65265f#diff-77f4caff2f2cfd8a790d6350d1cfc575a33cd96b81659f3acef8dc4d9be36b67L256

File: liberum_core/src/swarm_runner/behaviour/kademlia.rs [hunk 20]
```diff
@@ -1,87 +1,353 @@
...
+        if let Some(sender) = self.behaviour.pending_get_providers.remove(&id) {
+            let _ = sender.send(providers).inspect_err(|e| {
+                debug!(
+                    node = self.node.name,
+                    qid = format!("{id}"),
+                    err = format!("{e:?}"),
+                    "Channel closed"
+                )
+            });
+            // Finish the query to prevent from triggering FinishedWithNoAdditionalRecord
```
[the remark this answers is not in the document]
Comment (philipandag) [reply liberum_core/src/swarm_runner/behaviour/kademlia.rs:202]: I don't fully understand how this query progresses yet. It doesn't seem to be well documented. The example usage from libp2p-rust repo finishes the query here. The events structure suggests that `QueryResult::GetProviders` means the query was finished as it contains a `GetProvidersResult` which is a Result, of either `GetProvidersOk::FinishedWithNoAdditionalRecord` or `GetProvidersOk::FoundProviders`.

I think when writing the comment I understood that the query would continue as long as it finds any new provider and stop with `FinishedWithNoAdditionalRecord` when no more is found. The finish there is called because the `FinishedWithNoAdditionalRecord` is also handled and it too sends a response to the node calling `GetProviders`. That's the way they demonstrated the usage of GetProviders in libp2p-rust file-sharing example.

How to best handle this request is not apparent to me and I haven't found anything useful in the documentation. Maybe analysing more examples would help. Either way it's working and fixing it now would probably take too much time. Let's add an issue for that ok?

[the remark this answers is not in the document]
Comment (philipandag) [reply liberum_core/src/swarm_runner/behaviour/kademlia.rs:202]: https://github.com/Liberum-Neto/liberum-neto/issues/49

File: liberum_core/src/swarm_runner/behaviour/kademlia.rs [hunk 21]
```diff
@@ -1,87 +1,353 @@
...
+                id,
+                closest_peers,
+                stats,
+                step,
+            ),
+
+            // Triggered when the node is asked to put a record into it's store as a part of the DHT
+            Event::InboundRequest {
+                request:
+                    InboundRequest::PutRecord {
```
[the remark this answers is not in the document]
Comment (philipandag) [reply liberum_core/src/swarm_runner/behaviour/kademlia.rs]: `PutRecord` is an enum, not a struct :)
I have split the large kad::Event match into two smaller ones https://github.com/Liberum-Neto/liberum-neto/pull/43/commits/693a96259125293968ea12ed6ee4336aae7ae0c7

File: liberum_core/src/connection/mod.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (originalmk) [reply liberum_core/src/connection/mod.rs:86]: I think that if we want it back it may be easily readded in the future :)

File: liberum_core/src/swarm_runner/behaviour/kademlia.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (originalmk) [reply liberum_core/src/swarm_runner/behaviour/kademlia.rs]: But this is query_progressed and not finished, no?

File: liberum_core/src/connection/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (originalmk) [reply liberum_core/src/connection/mod.rs:56]: I think that if we want it back it may be easily readded in the future :)

File: liberum_core/src/swarm_runner/behaviour/kademlia.rs (hunk 1 above)
Comment (originalmk) [reply liberum_core/src/swarm_runner/behaviour/kademlia.rs:265]: Let's do PR for that.

File: liberum_core/src/swarm_runner/behaviour/kademlia.rs (hunk 2 above)
Comment (originalmk) [reply liberum_core/src/swarm_runner/behaviour/kademlia.rs:279]: Are providers records treated like every other record?

File: liberum_core/src/swarm_runner/messages.rs (hunk 5 above)
Comment (originalmk) [reply liberum_core/src/swarm_runner/messages.rs:218]: In normal network it should probably be more, so that kad will try to send more and more requests until quorum is satisfied.

File: liberum_core/src/swarm_runner/behaviour/kademlia.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (philipandag) [reply liberum_core/src/swarm_runner/behaviour/kademlia.rs]: There is no query finished event. OutboundQueryProgressed is used to show anything related to these queries, including when they finish

File: liberum_core/src/connection/mod.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (philipandag) [reply liberum_core/src/connection/mod.rs]: https://github.com/Liberum-Neto/liberum-neto/compare/f18f9355e140be206b5c6c7e1880e2cd27af2440..1f2e3a8c9b7abd4ce7a7b36ba161ba05767ef534?expand=1

File: liberum_core/src/connection/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (philipandag) [reply liberum_core/src/connection/mod.rs]: https://github.com/Liberum-Neto/liberum-neto/compare/f18f9355e140be206b5c6c7e1880e2cd27af2440..1f2e3a8c9b7abd4ce7a7b36ba161ba05767ef534?expand=1

File: liberum_core/src/swarm_runner/behaviour/kademlia.rs (hunk 2 above)
Comment (philipandag) [reply liberum_core/src/swarm_runner/behaviour/kademlia.rs:279]: Provider records and the actual data records are two seperate things in kad. They just decided to use the size of the K-bucket as the maximum amount of providers of a given ID to be remembered.

**Review by originalmk — DISMISSED** [review]

**Comment by originalmk** [comment]
There are still some warnings when building, could you fix these @philipandag?

**Review by originalmk — APPROVED** [review]
